### ISSUE TITLE
Why is there no support for directly computing cross entropy?

### ISSUE THREAD

**Issue by ushnish** [tensorflow/tensorflow#2462]
I see that we have methods for computing softmax and sigmoid cross entropy, which involve taking the softmax or sigmoid of the logit vector and then computing cross entropy with the target, and the weighted and sparse implementations of these. But what if I simply want to compute the cross entropy between 2 vectors?

**Comment by mrry** [comment]
We provide optimized cross-entropy implementations that are fused with the softmax/sigmoid implementations because their performance and numerical stability are critical to efficient training.

If however you are just interested in the cross entropy itself, you can compute it directly using code from the [beginners tutorial](https://www.tensorflow.org/versions/r0.8/tutorials/mnist/beginners/index.html):

```python
cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(y), reduction_indices=[1]))
```

**N.B.** DO NOT use this code for training. Use [`tf.nn.softmax_cross_entropy_with_logits()`](https://www.tensorflow.org/versions/r0.8/api_docs/python/nn.html#softmax_cross_entropy_with_logits) instead.

**Comment by stolsvik** [comment]
It would be great if the "numerical unstability" with the naive implementation was explained somewhere - I am trying to learn this stuff, and I wonder what that implies.

**Comment by b4be1** [comment]
@stolsvik Take a look at [this post](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html) to get some insight

**Comment by petrux** [comment]
I found this implementation from the [Keras code](https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py#L2717). As far as I could get, they just clip the `output` (i.e. the softmax) between `EPSILON` and `1-EPSILON`, perform a `log` operation on it and us the result as the logits. Could make sense?  
  
```python
def sparse_categorical_crossentropy(output, target, from_logits=False):
    """Categorical crossentropy with integer targets.
    # Arguments
        output: A tensor resulting from a softmax
            (unless `from_logits` is True, in which
            case `output` is expected to be the logits).
        target: An integer tensor.
        from_logits: Boolean, whether `output` is the
            result of a softmax, or is a tensor of logits.
    # Returns
        Output tensor.
    """
    # Note: tf.nn.softmax_cross_entropy_with_logits
    # expects logits, Keras expects probabilities.
    if not from_logits:
        epsilon = _to_tensor(_EPSILON, output.dtype.base_dtype)
        output = tf.clip_by_value(output, epsilon, 1 - epsilon)
        output = tf.log(output)

    output_shape = output.get_shape()
    targets = cast(flatten(target), 'int64')
    logits = tf.reshape(output, [-1, int(output_shape[-1])])
    res = tf.nn.sparse_softmax_cross_entropy_with_logits(
        labels=targets,
        logits=logits)
    if len(output_shape) == 3:
        # if our output includes timesteps we need to reshape
        return tf.reshape(res, tf.shape(output)[:-1])
    else:
        return res
```

**Comment by zrion** [comment]
Then in case that my output of the last layer is softmax activated, I have to manually do the cross entropy?
Then I don't understand why it is not supported because many models pick the output based on probabilities from softmax activation.

**Comment by tengerye** [comment]
Is it possible to add the `cross_entropy_with_logits()`? `tf.nn.softmax_cross_entropy_with_logits()` can work for the performance, but it makes codes difficult to read and implement. Thank you in advance. @mrry

**Comment by BrazilForever11** [comment]
`cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(y), reduction_indices=[1]))`

return nan when predict class has probability zero

**Comment by mrgloom** [comment]
Is clipping with `eps` still not make solution stable?

Example:

~~~
# Array of samples
y_pred_arr = np.array([[0.4,0.3,0.3], [1.0,0.0,0.0], [0.0,0.0,1.0]], np.float32)
y_true_arr = np.array([[1,0,0], [1,0,0], [0,0,1]], np.float32)

print('y_pred_arr.shape', y_pred_arr.shape)
print('y_true_arr.shape', y_true_arr.shape)

# 1 : Implemented in numpy
def cross_entropy_loss(y_pred_arr, y_true_arr, eps=1e-6):
    p_arr = np.clip(y_pred_arr, eps, 1 - eps)
    n_samples = p_arr.shape[0]
    n_classes = p_arr.shape[1]
    loss_arr = np.zeros((n_samples,), np.float32)
    for i in range(n_samples):
        for c in range(n_classes):
            loss_arr[i] += -y_true_arr[i][c] * np.log(p_arr[i][c])
    loss = np.mean(loss_arr)
    return loss
loss = cross_entropy_loss(y_pred_arr, y_true_arr)
print(loss)

# 2 : Implemented in sklearn
from sklearn.metrics import log_loss
loss = log_loss(np.argmax(y_true_arr,axis=1), y_pred_arr, labels=[0,1,2], eps=1e-6)
print(loss)

# 3 : Implemented in tensorflow
y_pred_tf = tf.convert_to_tensor(y_pred_arr, np.float32)
y_true_tf = tf.convert_to_tensor(y_true_arr, np.float32)
eps = 1e-6
cliped_y_pref_tf = tf.clip_by_value(y_pred_tf, eps, 1-eps)
loss_tf = tf.reduce_mean(-tf.reduce_sum(y_true_tf * tf.log(cliped_y_pref_tf), axis=1))
with tf.Session() as sess:
    loss = sess.run(loss_tf)
    print(loss)
~~~

Output:
~~~
y_pred_arr.shape (3, 3)
y_true_arr.shape (3, 3)
0.30543092
0.30543154478209544
0.30543092
~~~

**Comment by jondeaton** [comment]
Just wanted to add that a better implementation might be

```python
cross_entropy = tf.reduce_mean(-tf.reduce_sum(tf.math.xlogy(y_,  y), reduction_indices=[1]))
```

which will give `0` instead of `nans` where both `y_` and `y` are zero.
